torch/_inductor/fx_passes/fuse_attention.py

```diff
@@ -449,6 +473,86 @@ def _sfdp_replacement_17(query, key, value, attn_mask, inv_scale, dropout_p):
     )
 
 
+def _sfdp_pattern_18(query, key, value, inv_scale, causal_mask_value, causal_mask):
+    # for hf_GPT2 with dropout
+    query = query.permute([0, 2, 1, 3])
+    key = key.permute([0, 2, 1, 3])
+    value = value.permute([0, 2, 1, 3])
+    attn_weights = torch.matmul(query, key.permute(0, 1, 3, 2))
+    attn_weights = attn_weights.div(inv_scale)
+    attn_weights = torch.where(causal_mask, attn_weights, causal_mask_value)
+    return (
+        (
+            torch.nn.functional.dropout(attn_weights.softmax(dim=-1), 0.0)
+            .matmul(value)
+            .permute([0, 2, 1, 3])
+            .contiguous()
+        ),
+        key,
+        value,
+    )
+
+
+def _sfdp_replacement_18(query, key, value, inv_scale, causal_mask_value, causal_mask):
+    counters["inductor"]["fuse_attention"] += 1
+    query_permuted = query.permute([0, 2, 1, 3])
+    key_permuted = key.permute([0, 2, 1, 3])
+    value_permuted = value.permute([0, 2, 1, 3])
+    attn_bias = torch.zeros(1, 1, query.size(1), query.size(1))
+    return (
+        torch.ops.mkldnn._graph_sdpa_pattern(
```

Comment:
Could we map to this to `aten.scaled_dot_product_attention` then change the CPU implementation of `aten.scaled_dot_product_attention` to call oneDNN?

If not, what are the missing features?  I would like to have a common op shared between CPU/GPU rather than adding a CPU-specific op.

Reply:
If we'd invoke the oneDNN Graph kernel from `aten.scaled_dot_product_attention`, that'd result in regression for eager-mode because Graph Compiler backend of oneDNN Graph currently has higher compilation latency than the oneDNN kernels currently being used in PyTorch. :(
That's why I'm using a new op here, which wouldn't be exposed to users.

Reply:
Can we change the behavior of `aten.scaled_dot_product_attention` based on a global flag or environment variable?

These two patterns you added are CPU-only and won't work on GPU.  Also, after this PR we will have 19 attention patterns.  The other 17 attention patterns won't get mapped to oneDNN and be GPU-only.

We should have just one op that supports both CPU and GPU.  That way:
1) The 2 new patterns in the PR will work with CUDA
2) The 17 existing patterns will map to oneDNN when a mode is enabled

In the C++ version of scaled_dot_product_attention you can do:
```
if(global_flag) {
   return call_oneDNN_op(....)
} else {
  ... existing code with no compiles ...
}
```

Reply:
Hi @jansel, thanks again for reviewing!

It seems that following through on it would require modifying the semantics of the SDPA kernel as well, as well as its implementation, because it seems that the algorithm of new attention patterns being added in this PR is currently not supported by the CPU ATen SDPA kernel (based on the documentation of [`torch.nn.functional.scaled_dot_product_attention`](https://pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html)), so I'm not sure if they're supported by the CUDA SDPA kernel either.

After revision of the semantics of the ATen SDPA kernel as well as their corresponding implementation, I could add a flag in the C++ ATen CPU SDPA kernel that could call the oneDNN SDPA fusion op if the value of the environment variable `TORCHINDUCTOR_ONEDNN_GRAPH` would be `1`. But doing so doesn't seem possible right now.

Please advise if there's a potential workaround. Thanks!

Reply:
As for letting the existing patterns use oneDNN, I could follow through on this task in a subsequent PR - as you advised, in the ATen SDPA C++ kernel, I could check for the value of some environment variable, and then decide whether or not to use oneDNN.

However, the two new patterns would require using this new op.

Reply:
@jansel, @jgong5 had shared the same concern but since the ATen SDPA op does not support SDPA computation for GPT2 (the patterns in this PR), please advise how the new op can be used as a (conditional) fast-path in ATen code, since doing so requires changing the semantics of the ATen SDPA op.

Thanks!

Reply:
@jgong5, to summarize, in order to call the new (non-user-facing) op from the ATen SDPA C++ kernel for CPU, there are two blockers -

1. Modification of Semantics of `torch.nn.functional.scaled_dot_product_attention` and ATen SDPA kernels for CPU & CUDA, as well as
2.  Modification of ATen SDPA kernels for CPU & CUDA - someone would have to modify the CPU kernel, and someone would have to modify the CUDA kernel.

However, my concern is twofold:
1. Why would you like to block this implementation by doing those time-intensive things first?
Can they not be done down the line, and then the best performing implementation can be chosen?
2. Graph Compiler supports a lot of MHA/SDPA patterns (it will support more once some bugs are fixed in oneDNN v3.5). Right now, you're recommending adding GPT2 MHA/SDPA support in ATen SDPA kernel.
 Would you recommend modifying the ATen SDPA implementation each time it wouldn't be able to handle some model's MHA/SDPA implementation? For example, [this one](https://github.com/oneapi-src/oneDNN/blob/main/src/graph/backend/graph_compiler/patterns/mha_pattern.hpp#L860-L879)?

Please clarify, so that I may have clarity on any potential action-items. Thanks!

Reply:
I don't object to extending the SDPA to support GPT2 later. A cleaner approach in my mind is that we have a consistent way of matching these SDPA patterns to the SDPA ATen op/kernel, serving both CPU and CUDA as a baseline (currently we have 17 patterns working like this way and the additional 2 for GPT2 can be implemented similarly). Then, we have a fast path with oneDNN graph on the patterns that it can bring better perf. This can be implemented as a separate pass, running before the baseline SDPA pass. It should reuse the existing SDPA patterns but provide its own replacement functions.

Reply:
> This can be implemented as a separate pass, running before the baseline SDPA pass. It should reuse the existing SDPA patterns but provide its own replacement functions

Thanks, I also like this idea because it allows replacing one of the 17 existing patterns with oneDNN Graph replacements.

But could you please clarify how such a pass can share the two new patterns pertaining to GPT2, since those are unsupported by ATen SDPA ops of CPU & CUDA? Specifically, the constraint `It should reuse the existing SDPA patterns` can't be satisfied for GPT2 MHA/SDPA patterns simply because ATen CPU/CUDA SDPA ops don't support them.

Reply:
Hi @eelison @jansel, 2 weeks back, I had tried to add a pass that Jiong recommended in his [previous comment](https://github.com/pytorch/pytorch/pull/120301#discussion_r1517482275) - essentially, running a oneDNN Graph specific `sfdp_init` pass before the existing `sfdp_init` pass.
However, I ran into asserts that patterns had already been "seen", although they had different replacements.
Could you please suggest a workaround? Thanks!

Reply:
Hi @jansel @eellison, may I use two `PatternMatcherPass`es in `joint_graph_passes`, as is being done at some other places in the codebase (I see three being used at another place)? Thanks!

Reply:
I think we can extend the sdpa pattern in core to support GPT2.  Maybe we need a new function, or a new optional arg on the existing one.  @Chillee has also been talking about a more configurable attention API.  

Whatever we do it will (eventually) need to support both CPU/CUDA and that support should happen through a shared aten API.

cc @drisspg